### PR TITLE
In Source Order

### PR DESCRIPTION
**Pull request by j-baker**
This PR does a couple of things, one very minor.

We make it so that a `Stream<SettableFuture>` can be passed in, and will be returned as a `Stream<SettableFuture>`. This should be both binary and source backcompatible (and this release has only been around for a day or two anyway).

We also add a new `inSourceOrder` method which returns the futures in source order; basically the properties are exactly the same except each future will be blocked from being returned to the user by the previous ones.

Also add some more stringent tests to MoreStreams to ensure the correct methods are being delegated to.

### CODE REVIEW

File: src/test/java/com/palantir/common/streams/MoreFuturesTests.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    public void testBlockOnCompletion_cancelled() {
+        Future<String> future = Futures.immediateCancelledFuture();
+        MoreFutures.blockOnCompletion(future);
+        assertThat(future.isCancelled()).isTrue();
+    }
+
+    @Test
+    public void testBlockOnCompletion_blocking() throws InterruptedException, ExecutionException {
```
Comment (j-baker) [inline src/test/java/com/palantir/common/streams/MoreFuturesTests.java]: not sure this class needs to exist - the method is very very simple. This test though is specifically hard to write properly - it seems basically impossible to write a deterministic test for blocking.

File: src/main/java/com/palantir/common/streams/MoreStreams.java [hunk 2]
```diff
@@ -57,6 +59,30 @@
     }
 
     /**
+     * Given a {@code Stream<ListenableFuture<U>>}, this function will return a blocking stream of the completed
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreStreams.java]: this function will return a blocking stream that waits for each future to complete before returning it, but which looks ahead `maxParallelism` futures to ensure a fixed parallelism rate

File: src/main/java/com/palantir/common/streams/MoreStreams.java [hunk 3]
```diff
@@ -38,10 +40,10 @@
      * Given a {@code Stream<ListenableFuture<U>>}, this function will return a blocking stream of the completed
      * futures in completion order, looking at most {@code maxParallelism} futures ahead in the stream.
      */
-    public static <U> Stream<ListenableFuture<U>> inCompletionOrder(
-            Stream<ListenableFuture<U>> futures, int maxParallelism) {
-        return StreamSupport.stream(
-                new InCompletionOrderSpliterator<>(futures.spliterator(), maxParallelism), NOT_PARALLEL);
+    public static <T extends ListenableFuture<U>, U> Stream<T> inCompletionOrder(
+            Stream<T> futures, int maxParallelism) {
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreStreams.java]: just accept a `Stream<? extends ListenableFuture<U>>`

File: src/main/java/com/palantir/common/streams/MoreFutures.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.common.util.concurrent.Futures;
+
+import java.util.concurrent.Future;
+
+class MoreFutures {
+    private MoreFutures() {}
+
+    static <T extends Future<U>, U> T blockOnCompletion(T future) {
+        try {
+            Futures.getUnchecked(future);
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreFutures.java]: Better to use `future.get()` and explicitly handle the checked exceptions it throws, this way Errors don't trip you up and interrupted exceptions don't go missing.

File: src/main/java/com/palantir/common/streams/MoreFutures.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+package com.palantir.common.streams;
+
+import com.google.common.util.concurrent.Futures;
+
+import java.util.concurrent.Future;
+
+class MoreFutures {
+    private MoreFutures() {}
+
+    static <T extends Future<U>, U> T blockOnCompletion(T future) {
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreFutures.java]: `join` in line with J8 api?

File: src/main/java/com/palantir/common/streams/BufferingSpliterator.java [hunk 6]
```diff
@@ -88,4 +91,27 @@ public long estimateSize() {
     public int characteristics() {
         return Spliterator.SIZED & notStarted.characteristics();
     }
+
+    @FunctionalInterface
+    interface CompletionStrategy {
+        <T extends ListenableFuture<U>, U> void registerCompletion(T future, Consumer<T> resultConsumer);
+    }
+
+    enum InCompletionOrder implements CompletionStrategy {
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/BufferingSpliterator.java]: These can be a single enum

File: src/main/java/com/palantir/common/streams/MoreStreams.java [hunk 7]
```diff
@@ -57,6 +59,30 @@
     }
 
     /**
+     * Given a {@code Stream<ListenableFuture<U>>}, this function will return a blocking stream of the completed
+     * futures in the same order as the source, looking at most {@code maxParallelism} futures ahead in the stream.
+     */
+    public static <T extends ListenableFuture<U>, U> Stream<T> withParallelLookahead(
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreStreams.java]: After honing the JavaDoc, I think `blockingStreamWithParallelism` works better

File: src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java [hunk 8]
```diff
@@ -34,7 +35,10 @@
 import org.mockito.junit.MockitoJUnitRunner;
 
 @RunWith(MockitoJUnitRunner.class)
-public final class InCompletionOrderSpliteratorTests {
+public final class BufferingSpliteratorTests {
+    // None of the tests in this class should be affected by the choice of completion strategy.
```
Comment (chrisalice) [inline src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java]: Why not use a parameterized runner then?

Comment (chrisalice) [reply src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java]: Also, we really should have tests that distinguish the two cases

File: src/test/java/com/palantir/common/streams/MoreFuturesTests.java (hunk 1 above)
Comment (chrisalice) [reply src/test/java/com/palantir/common/streams/MoreFuturesTests.java]: Agreed. The class should probably exist, but this test doesn't need to.

File: src/main/java/com/palantir/common/streams/MoreStreams.java [hunk 9]
```diff
@@ -38,10 +40,10 @@
      * Given a {@code Stream<ListenableFuture<U>>}, this function will return a blocking stream of the completed
      * futures in completion order, looking at most {@code maxParallelism} futures ahead in the stream.
      */
-    public static <U> Stream<ListenableFuture<U>> inCompletionOrder(
-            Stream<ListenableFuture<U>> futures, int maxParallelism) {
-        return StreamSupport.stream(
-                new InCompletionOrderSpliterator<>(futures.spliterator(), maxParallelism), NOT_PARALLEL);
+    public static <T extends ListenableFuture<U>, U> Stream<T> inCompletionOrder(
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreStreams.java]: U first?

File: src/main/java/com/palantir/common/streams/MoreStreams.java [hunk 10]
```diff
@@ -38,10 +40,10 @@
      * Given a {@code Stream<ListenableFuture<U>>}, this function will return a blocking stream of the completed
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreStreams.java]: "Given a stream of listenable futures"...

File: src/main/java/com/palantir/common/streams/MoreFutures.java (hunk 5 above)
Comment (j-baker) [reply src/main/java/com/palantir/common/streams/MoreFutures.java]: no, that has different semantics. I don't want this to throw, whereas join() will throw.

File: src/main/java/com/palantir/common/streams/MoreFutures.java (hunk 4 above)
Comment (j-baker) [reply src/main/java/com/palantir/common/streams/MoreFutures.java]: sure.

File: src/main/java/com/palantir/common/streams/BufferingSpliterator.java (hunk 6 above)
Comment (j-baker) [reply src/main/java/com/palantir/common/streams/BufferingSpliterator.java]: what do you mean?

File: src/test/java/com/palantir/common/streams/MoreFuturesTests.java (hunk 1 above)
Comment (j-baker) [reply src/test/java/com/palantir/common/streams/MoreFuturesTests.java]: yes, but this function is the only one that actually would actually fail if you just didn't call the function.

Comment (j-baker) [reply src/test/java/com/palantir/common/streams/MoreFuturesTests.java]: so basically i'm thinking all or none?

File: src/main/java/com/palantir/common/streams/MoreStreams.java (hunk 2 above)
Comment (j-baker) [reply src/main/java/com/palantir/common/streams/MoreStreams.java]: sure

File: src/main/java/com/palantir/common/streams/MoreStreams.java (hunk 7 above)
Comment (j-baker) [reply src/main/java/com/palantir/common/streams/MoreStreams.java]: done

File: src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java (hunk 8 above)
Comment (j-baker) [reply src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java]: We do have such tests - they're in MoreFuturesTests.

I'll do the parameterization.

File: src/test/java/com/palantir/common/streams/MoreFuturesTests.java (hunk 1 above)
Comment (chrisalice) [reply src/test/java/com/palantir/common/streams/MoreFuturesTests.java]: Oh... yeah, okay, kill it.

File: README.md [hunk 11]
```diff
@@ -56,6 +56,10 @@ executor, calling
 
     MoreStreams.inCompletionOrder(foos.stream(), service::getBar, executor, maxParallelism).collect(toList());
 
+The difference between `inCompletionOrder` and `blockingStreamWithParallelism` is that the `blockStreamWithParallelism`
+methods keep the futures in the order they were provided, whilst the `inCompletionOrder` methods reorder the stream
+into completion order.
```
Comment (chrisalice) [inline README.md]: return futures as they complete

File: src/main/java/com/palantir/common/streams/BufferingSpliterator.java (hunk 6 above)
Comment (chrisalice) [reply src/main/java/com/palantir/common/streams/BufferingSpliterator.java]:
```
enum CompletionStrategy {
  SOURCE_ORDER() {
    ...
  },
  COMPLETION_ORDER() {
    ...
  };
  ...
}
```

File: src/main/java/com/palantir/common/streams/MoreFutures.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+class MoreFutures {
+    private MoreFutures() {}
+
+    static <T extends Future<U>, U> T blockUntilCompletion(T future) {
+        try {
+            future.get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        } catch (ExecutionException | RuntimeException | Error ignored) {}
```
Comment (chrisalice) [inline src/main/java/com/palantir/common/streams/MoreFutures.java]: Remove RuntimeException and Error?

File: src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java [hunk 13]
```diff
@@ -23,27 +23,47 @@
...
+                new Object[] {"InCompletionOrder", InCompletionOrder.INSTANCE});
+    }
+
+    public BufferingSpliteratorTests(String name, CompletionStrategy completionStrategy) {
+        this.completionStrategy = completionStrategy;
+    }
+
     @Before
     public void before() {
+        MockitoAnnotations.initMocks(this);
```
Comment (chrisalice) [inline src/test/java/com/palantir/common/streams/BufferingSpliteratorTests.java]: Use MockitoRule, it has a useful teardown check

**Pull request merged**